### PR TITLE
【Hackathon4 No.61】remainder 算子FP16/BF16单测完善

### PR DESCRIPTION
**Pull request by enkilee**
### PR types

Others
### PR changes

Others
### Description

remainder

### CODE REVIEW

**Comment by enkilee** [comment]
单测框架能自动为fp16跳过不支持的设备，可是
2023-04-20 16:29:57 1/2 Test  #376: test_elementwise_mod_op ..........***Failed    2.04 sec
2023-04-20 16:29:57 Hint: Your machine support AVX, but the installed paddlepaddle doesn't have avx core. Hence, no-avx core with worse performance will be imported.
2023-04-20 16:29:57 If you like, you could reinstall paddlepaddle by 'python -m pip install --force-reinstall paddlepaddle-gpu[==version]' to get better performance.
2023-04-20 16:29:57 E0420 08:29:57.602474 378957 place.cc:349] Cannot use GPU because you have installed CPU version PaddlePaddle.
2023-04-20 16:29:57 If you want to use GPU, please try to install GPU version PaddlePaddle by: pip install paddlepaddle-gpu
2023-04-20 16:29:57 If you only have CPU, please change CUDAPlace(0) to be CPUPlace().

**Comment by enkilee** [comment]
2023-05-04 22:39:21 Not equal to tolerance rtol=0.05, atol=0
2023-05-04 22:39:21 Mismatched elements: 3 / 10 (30%)
2023-05-04 22:39:21 Max absolute difference: 0.03
2023-05-04 22:39:21 Max relative difference: 1.
2023-05-04 22:39:21  x: array([0.  , 0.  , 0.  , 0.03, 0.  , 0.01, 0.  , 0.  , 0.  , 0.  ],
2023-05-04 22:39:21       dtype=float32)
2023-05-04 22:39:21  y: array([0.  , 0.  , 0.  , 0.  , 0.02, 0.01, 0.  , 0.01, 0.  , 0.  ],
2023-05-04 22:39:21       dtype=float32)

按照这个误差来？

File: python/paddle/fluid/tests/unittests/test_elementwise_mod_op.py [hunk 1]
```diff
@@ -106,14 +111,17 @@ def test_check_output(self):
...
+)
+class TestElementwiseModFP16Op(TestElementwiseModOp):
     def init_dtype(self):
         self.dtype = np.float16
 
     def init_input_output(self):
         self.x = np.random.uniform(-1000, 1000, [10, 10]).astype(self.dtype)
         self.y = np.random.uniform(-100, 100, [10, 10]).astype(self.dtype)
-        self.out = np.mod(self.x, self.y)
+        self.out = np.fmod(self.y + np.fmod(self.x, self.y), self.y)
```
Comment (ZzSean) [inline python/paddle/fluid/tests/unittests/test_elementwise_mod_op.py:124]: 这么改的目的是什么

File: python/paddle/fluid/tests/unittests/test_elementwise_mod_op.py [hunk 2]
```diff
@@ -122,6 +130,97 @@ def test_check_output(self):
...
+        self.attrs = {'axis': self.axis, 'use_mkldnn': self.use_mkldnn}
+        self.outputs = {'Out': convert_float_to_uint16(self.out)}
+
+    def test_check_output(self):
+        place = core.CUDAPlace(0)
+        self.check_output_with_place_customized(
+            checker=self.verify_output, place=place
+        )
+
+    def verify_output(self, outs):
```
Comment (ZzSean) [inline python/paddle/fluid/tests/unittests/test_elementwise_mod_op.py]: 为什么需要自定义一个计算函数

File: python/paddle/fluid/tests/unittests/test_elementwise_mod_op.py (hunk 1 above)
Comment (enkilee) [reply python/paddle/fluid/tests/unittests/test_elementwise_mod_op.py:124]: 上面的float是用的这种

File: python/paddle/fluid/tests/unittests/test_elementwise_mod_op.py (hunk 2 above)
Comment (enkilee) [reply python/paddle/fluid/tests/unittests/test_elementwise_mod_op.py]: 不管是用np.mod还是np.fmod，rtol都到了300以上

Comment (ZzSean) [reply python/paddle/fluid/tests/unittests/test_elementwise_mod_op.py]: 由于数据类型所限，使用fp32计算和使用bf16的结果差距太大，这里可以对输入做一些处理，初始化为FP32->convert为bf16->convert为fp32，即`convert_uint16_to_float(convert_float_to_uint16(x))`然后再计算参考结果，这样应该是可以通过的，无需使用自定义函数

Comment (enkilee) [reply python/paddle/fluid/tests/unittests/test_elementwise_mod_op.py]: 2023-05-11 16:06:08 AssertionError: 
2023-05-11 16:06:08 Not equal to tolerance rtol=1e-05, atol=0.01
2023-05-11 16:06:08 Operator (elementwise_mod) Output (Out) has diff at Place(cpu) in static checker
2023-05-11 16:06:08 Mismatched elements: 90 / 100 (90%)
2023-05-11 16:06:08 Max absolute difference: 394.
2023-05-11 16:06:08 Max relative difference: 10.13364
2023-05-11 16:06:08  x: array([[288.     , 192.     ,  40.5    ,  44.     , 128.     , 264.     ,
2023-05-11 16:06:08         508.     , 224.     , 540.     ,  80.     ],
2023-05-11 16:06:08        [264.     , 184.     , 109.     , 596.     , 250.     , 184.     ,...
2023-05-11 16:06:08  y: array([[294.      , 194.      ,  44.25    ,  15.5     , 111.      ,
2023-05-11 16:06:08         260.      , 496.      , 209.      , 536.      ,  72.5     ],
2023-05-11 16:06:08        [262.      , 205.      , 108.5     , 596.      , 255.      ,...
2023-05-11 16:06:08 ----------------------------------------------------------------------

Comment (enkilee) [reply python/paddle/fluid/tests/unittests/test_elementwise_mod_op.py]: 政哥，我试了，还是不行，初始化为FP32->convert为bf16->convert为fp32后，仍然rtol到了300多。

Comment (enkilee) [reply python/paddle/fluid/tests/unittests/test_elementwise_mod_op.py]: 2023-05-13 18:23:41 Not equal to tolerance rtol=1e-05, atol=0.01
2023-05-13 18:23:41 Operator (elementwise_mod) Output (Out) has diff at Place(gpu:0) in static checker
2023-05-13 18:23:41 Mismatched elements: 90 / 100 (90%)
2023-05-13 18:23:41 Max absolute difference: 394.
2023-05-13 18:23:41 Max relative difference: 10.13364
2023-05-13 18:23:41  x: array([[288.     , 192.     ,  40.5    ,  44.     , 128.     , 264.     ,
2023-05-13 18:23:41         508.     , 224.     , 540.     ,  80.     ],
2023-05-13 18:23:41        [264.     , 184.     , 109.     , 596.     , 250.     , 184.     ,...
2023-05-13 18:23:41  y: array([[294.      , 194.      ,  44.25    ,  15.5     , 111.      ,
2023-05-13 18:23:41         260.      , 496.      , 209.      , 536.      ,  72.5     ],
2023-05-13 18:23:41        [262.      , 205.      , 108.5     , 596.      , 255.      ,...

**Review by ZzSean — APPROVED** [review]
LGTM

**Pull request merged**